framework/src/base/FEProblem.C

```diff
@@ -555,13 +560,23 @@ void FEProblem::initialSetup()
       computeUserObjects(EXEC_TIMESTEP_BEGIN);
       computeUserObjects(EXEC_LINEAR);
     }
-    Moose::setup_perf_log.pop("Initial computeUserObjects()","Setup");
-  }
 
-  _nl.initialSetupBCs();
-  _nl.initialSetupKernels();
+    // Perform an initial residual evaluation to set save-in variables
+    // Note: time kernels need not to be evaluated on initial because their contribution should be just zero.
+    //       They actually cannot be evaluated because we set the time step size to zero on initial.
+    if (_nl.hasSaveIn())
+    {
+      _nl.computeResidual(*_nl.sys().rhs, Moose::KT_NONTIME);
+      // We want to call post user object evaluation one more time in case they depends on the save-in variables.
+      computeUserObjects(EXEC_INITIAL, UserObjectWarehouse::POST_AUX);
```

Comment:
I think most of the changes to NonlinearSystem are good.  They are nice and generic and will give us more information about what's going on.